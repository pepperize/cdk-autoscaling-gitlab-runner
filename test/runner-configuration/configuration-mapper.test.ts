import { AnyJson } from "@iarna/toml";
import { ConfigurationMapper, ConfigurationMapperProps, filter, isEmpty } from "../../src";

describe("ConfigurationMapper", () => {
  it("Should return empty map when provided a configuration consisting of empty values", () => {
    // Given
    const props: ConfigurationMapperProps = {
      globalConfiguration: {},
      runnersConfiguration: [
        {
          docker: {
            allowedImages: [],
          },
          machine: {
            autoscaling: [],
          },
          cache: {
            s3: {},
          },
        },
      ],
    };

    // When
    const mapper = ConfigurationMapper.fromProps(props);
    const actual = mapper._toJsonMap();

    // Then
    expect(actual).toEqual({}); // Should be empty
  });

  it("Should return empty map when provided a configuration consisting of undefined and empty values", () => {
    // Given
    const props: ConfigurationMapperProps = {
      globalConfiguration: {
        logFormat: undefined,
      },
      runnersConfiguration: [
        {
          docker: {},
          machine: {
            machineOptions: {
              blockDurationMinutes: undefined,
            },
            autoscaling: [],
          },
          cache: {
            s3: {},
          },
        },
      ],
    };

    // When
    const mapper = ConfigurationMapper.fromProps(props);
    const actual = mapper._toJsonMap();

    // Then
    expect(actual).toEqual({}); // Should be empty
  });

  it("Should return map with defaults when provided a configuration consisting of empty values", () => {
    // Given
    const props: ConfigurationMapperProps = {
      globalConfiguration: {},
      runnersConfiguration: [
        {
          docker: {},
          machine: {},
          cache: {
            s3: {},
          },
        },
      ],
    };

    // When
    const mapper = ConfigurationMapper.withDefaults(props);
    const actual = mapper._toJsonMap();

    // Then
    expect(actual).toEqual({
      check_interval: 0,
      concurrent: 10,
      log_format: "runner",
      log_level: "info",
      runners: [
        {
          environment: [
            "DOCKER_DRIVER=overlay2",
            "DOCKER_TLS_CERTDIR=/certs",
            'DOCKER_AUTH_CONFIG={"credsStore": "ecr-login"}',
          ],
          executor: "docker+machine",
          limit: 10,
          output_limit: 52428800,
          url: "https://gitlab.com",
          docker: {
            tls_verify: false,
            image: "docker:19.03.5",
            privileged: true,
            cap_add: ["CAP_SYS_ADMIN"],
            wait_for_services_timeout: 300,
            disable_cache: false,
            volumes: ["/certs/client", "/cache"],
            shm_size: 0,
          },
          machine: {
            IdleCount: 0,
            IdleTime: 300,
            MaxBuilds: 20,
            MachineDriver: "amazonec2",
            MachineName: "gitlab-runner-%s",
            MachineOptions: [
              "amazonec2-request-spot-instance=true",
              "amazonec2-spot-price=0.03",
              "amazonec2-metadata-token=required",
              "amazonec2-metadata-token-response-hop-limit=2",
            ],
            autoscaling: [{ Periods: ["* * 7-22 * * mon-fri *"], IdleCount: 1, IdleTime: 1800, Timezone: "Etc/UTC" }],
          },
        },
      ],
    });
  });
  it("Should map snapshot with defaults only at empty values when provided a configuration consisting of a mix of empty and defined values", () => {
    // Given
    const props: ConfigurationMapperProps = {
      globalConfiguration: {},
      runnersConfiguration: [
        {
          token: "foo+bar",
          docker: {},
          machine: {},
          cache: {
            s3: {
              serverAddress: "s3.amazonaws.com",
              bucketName: "gitlab-runner-cache-bucket-test-us-east-1",
              bucketLocation: "us-east-1",
              accessKey: "AKIAIOSFODNN7EXAMPLE",
              secretKey: "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
            },
          },
        },
      ],
    };

    // When
    const mapper = ConfigurationMapper.withDefaults(props);
    const actual = mapper.toToml();

    // Then
    expect(actual).toMatchSnapshot();
  });
  it("Should return empty values when provided empty values", () => {
    // Given
    const props: AnyJson = {
      array: [{ empty: undefined }],
    } as unknown as AnyJson;

    // When
    const actual = filter(props, (item) => !isEmpty(item));

    // Then
    expect(actual).toEqual({});
  });
  it("Should match snapshot when provided multiple runners configuration", () => {
    // Given
    const props: ConfigurationMapperProps = {
      globalConfiguration: {},
      runnersConfiguration: [
        {
          token: "foo+bar",
          docker: {},
          machine: {},
          cache: {
            s3: {
              serverAddress: "s3.amazonaws.com",
              bucketName: "gitlab-runner-cache-bucket-test-us-east-1",
              bucketLocation: "us-east-1",
              accessKey: "AKIAIOSFODNN7EXAMPLE",
              secretKey: "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
            },
          },
        },
        {
          token: "2foo+bar",
          docker: {},
          machine: {
            autoscaling: [],
          },
          cache: {
            s3: {
              serverAddress: "2s3.amazonaws.com",
              bucketName: "2gitlab-runner-cache-bucket-test-us-east-1",
              bucketLocation: "2us-east-1",
              accessKey: "2AKIAIOSFODNN7EXAMPLE",
              secretKey: "2wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
            },
          },
        },
      ],
    };

    // When
    const mapper = ConfigurationMapper.withDefaults(props);
    const actual = mapper.toToml();

    // Then
    expect(actual).toMatchSnapshot();
  });
});
