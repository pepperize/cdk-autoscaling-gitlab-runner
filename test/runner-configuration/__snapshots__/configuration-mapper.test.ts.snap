// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigurationMapper Should map snapshot with defaults only at empty values when provided a configuration consisting of a mix of empty and defined values 1`] = `
"concurrent = 10
check_interval = 0
log_format = \\"runner\\"
log_level = \\"info\\"

[[runners]]
url = \\"https://gitlab.com\\"
limit = 10
output_limit = 52_428_800
executor = \\"docker+machine\\"
environment = [
  \\"DOCKER_DRIVER=overlay2\\",
  \\"DOCKER_TLS_CERTDIR=/certs\\",
  \\"DOCKER_AUTH_CONFIG={\\\\\\"credsStore\\\\\\": \\\\\\"ecr-login\\\\\\"}\\"
]
token = \\"foo+bar\\"

  [runners.docker]
  tls_verify = false
  image = \\"docker:19.03.5\\"
  privileged = true
  cap_add = [ \\"CAP_SYS_ADMIN\\" ]
  wait_for_services_timeout = 300
  disable_cache = false
  volumes = [ \\"/certs/client\\", \\"/cache\\" ]
  shm_size = 0

  [runners.machine]
  IdleCount = 0
  IdleTime = 300
  MaxBuilds = 20
  MachineDriver = \\"amazonec2\\"
  MachineName = \\"gitlab-runner-%s\\"
  MachineOptions = [
  \\"amazonec2-request-spot-instance=true\\",
  \\"amazonec2-spot-price=0.03\\",
  \\"amazonec2-metadata-token=required\\",
  \\"amazonec2-metadata-token-response-hop-limit=2\\"
]

    [[runners.machine.autoscaling]]
    Periods = [ \\"* * 7-22 * * mon-fri *\\" ]
    IdleCount = 1
    IdleTime = 1_800
    Timezone = \\"Etc/UTC\\"

  [runners.cache]
  Type = \\"s3\\"
  Shared = true

    [runners.cache.S3]
    serverAddress = \\"s3.amazonaws.com\\"
    bucketName = \\"gitlab-runner-cache-bucket-test-us-east-1\\"
    bucketLocation = \\"us-east-1\\"
    accessKey = \\"AKIAIOSFODNN7EXAMPLE\\"
    secretKey = \\"wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY\\"
"
`;

exports[`ConfigurationMapper Should match snapshot when provided multiple runners configuration 1`] = `
"concurrent = 10
check_interval = 0
log_format = \\"runner\\"
log_level = \\"info\\"

[[runners]]
url = \\"https://gitlab.com\\"
limit = 10
output_limit = 52_428_800
executor = \\"docker+machine\\"
environment = [
  \\"DOCKER_DRIVER=overlay2\\",
  \\"DOCKER_TLS_CERTDIR=/certs\\",
  \\"DOCKER_AUTH_CONFIG={\\\\\\"credsStore\\\\\\": \\\\\\"ecr-login\\\\\\"}\\"
]
token = \\"foo+bar\\"

  [runners.docker]
  tls_verify = false
  image = \\"docker:19.03.5\\"
  privileged = true
  cap_add = [ \\"CAP_SYS_ADMIN\\" ]
  wait_for_services_timeout = 300
  disable_cache = false
  volumes = [ \\"/certs/client\\", \\"/cache\\" ]
  shm_size = 0

  [runners.machine]
  IdleCount = 0
  IdleTime = 300
  MaxBuilds = 20
  MachineDriver = \\"amazonec2\\"
  MachineName = \\"gitlab-runner-%s\\"
  MachineOptions = [
  \\"amazonec2-request-spot-instance=true\\",
  \\"amazonec2-spot-price=0.03\\",
  \\"amazonec2-metadata-token=required\\",
  \\"amazonec2-metadata-token-response-hop-limit=2\\"
]

    [[runners.machine.autoscaling]]
    Periods = [ \\"* * 7-22 * * mon-fri *\\" ]
    IdleCount = 1
    IdleTime = 1_800
    Timezone = \\"Etc/UTC\\"

  [runners.cache]
  Type = \\"s3\\"
  Shared = true

    [runners.cache.S3]
    serverAddress = \\"s3.amazonaws.com\\"
    bucketName = \\"gitlab-runner-cache-bucket-test-us-east-1\\"
    bucketLocation = \\"us-east-1\\"
    accessKey = \\"AKIAIOSFODNN7EXAMPLE\\"
    secretKey = \\"wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY\\"

[[runners]]
url = \\"https://gitlab.com\\"
limit = 10
output_limit = 52_428_800
executor = \\"docker+machine\\"
environment = [
  \\"DOCKER_DRIVER=overlay2\\",
  \\"DOCKER_TLS_CERTDIR=/certs\\",
  \\"DOCKER_AUTH_CONFIG={\\\\\\"credsStore\\\\\\": \\\\\\"ecr-login\\\\\\"}\\"
]
token = \\"2foo+bar\\"

  [runners.docker]
  tls_verify = false
  image = \\"docker:19.03.5\\"
  privileged = true
  cap_add = [ \\"CAP_SYS_ADMIN\\" ]
  wait_for_services_timeout = 300
  disable_cache = false
  volumes = [ \\"/certs/client\\", \\"/cache\\" ]
  shm_size = 0

  [runners.machine]
  IdleCount = 0
  IdleTime = 300
  MaxBuilds = 20
  MachineDriver = \\"amazonec2\\"
  MachineName = \\"gitlab-runner-%s\\"
  MachineOptions = [
  \\"amazonec2-request-spot-instance=true\\",
  \\"amazonec2-spot-price=0.03\\",
  \\"amazonec2-metadata-token=required\\",
  \\"amazonec2-metadata-token-response-hop-limit=2\\"
]

  [runners.cache]
  Type = \\"s3\\"
  Shared = true

    [runners.cache.S3]
    serverAddress = \\"2s3.amazonaws.com\\"
    bucketName = \\"2gitlab-runner-cache-bucket-test-us-east-1\\"
    bucketLocation = \\"2us-east-1\\"
    accessKey = \\"2AKIAIOSFODNN7EXAMPLE\\"
    secretKey = \\"2wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY\\"
"
`;
